libs/core/src/operators/substream.pipe.ts

```diff
@@ -0,0 +1,106 @@
+    const destination = this.destination;
+
+    // call the predicate every time
+    // this is the only change from the vanilla rxjs operator
+    this.tryCallPredicate(value);
+
+    if (!this.skipping) {
+      destination.next(value);
+    }
+  }
+
+  private tryCallPredicate(value: T): void {
+    try {
+      const result = this.predicate(value, this.index++);
+      this.skipping = Boolean(result);
+    } catch (err) {
+      this.destination.error(err);
+    }
+  }
+}
+
+class SkipWhileOperator<T> implements Operator<T, T> {
+  constructor(private predicate: (value: T, index: number) => boolean) {
+  }
+
+  call(subscriber: Subscriber<T>, source: any): TeardownLogic {
+    return source.subscribe(new SkipWhileSubscriber(subscriber, this.predicate));
+  }
+}
+
+/**
+ * All the skipWhile stuff is copied directly from rxjs/src/internal/operators/skipWhile.ts.
+ * The only change is that it will continue to check the predicate even while not skipping.
+ */
+function skipWhile<T>(predicate: (value: T, index: number) => boolean): MonoTypeOperatorFunction<T> {
+  return (source: Observable<T>) => source.lift(new SkipWhileOperator(predicate));
+}
+
+export type ActionType = Action['type'];
+export type ActionSequenceStep = ActionType | ActionType[];
+export type ActionSequence = ActionSequenceStep[];
+
+const compareActionSequenceStep = (step: ActionSequenceStep, action: Action): boolean => Array.isArray(step)
+  ? step.reduce((acc, type) => acc || action.type === type, false)
+  : action.type === step
+
+/**
+ * Finds and emits a substream of actions.
+ *
+ * This will find the first occurence of the substream and not necessarily the most concise or recent substream.
+ * For example, substream([A, B]) encountering an action stream of [A1, A2, B1] will emit [A1, B1].
+ * When a substream is found, the list is reset and it will continue to listen to the action stream.
+ *
+ * You can indicate that a particular step in the sequence can be matched by mutiple actions by passing an array for that step.
+ * For example, if the first step of the sequence could be action A or action B: substream([[A, B], C]).
+ * An action stream of [A1, C1] would cause substream to emit [A1, C1].
+ * An action stream of [B1, C1] would cause substream to emit [B1, C1].
+ * An action stream of [A1, B1, C1] would cause substream to emit [A1, C1].
+ *
+ * You can optionally pass a list of terminators.
+ * If a terminator is encountered in the action stream, any partially matched substream is reset.
+ * For example, if substream([A, B], C), an action stream of [A1, C1, B1] would cause substream to not emit.
+ *
+ * @param sequence The sequence of action types that define the substream to listen for.
+ * @param terminators A list of terminators, any of which interrupt and reset a partially matched substream.
+ */
+export const substream = (sequence: ActionSequence, ...terminators: ActionType[]) => {
+  const buildMatchedSequence = (matchedSequence, action) =>
```

Comment:
I think this function should be named something else like "accumulateStream" or "accumulateSequence" or something like that.

Reply:
How about `accumulateMatchedStream`? yeah, you're right what I call the sequence and the stream are different data structures. I shouldn't mix the terms.

Reply:
Oh, good point. I was actually just talking about the word "build". `accumulateMatchedSubstream` sounds good

Reply:
oh right I don't think I read your comment fully. not sure what I was thinking. changed.